Support writing-mode conversions for position-area

fantasai opened the pull request:
#### 054862eb3b5d878ad8399239d949d60ff62cb5c9
<pre>
Support writing-mode conversions for position-area
<a href="https://bugs.webkit.org/show_bug.cgi?id=288261">https://bugs.webkit.org/show_bug.cgi?id=288261</a>
<a href="https://rdar.apple.com/145333147">rdar://145333147</a>

Reviewed by Alan Baradlay.

Adds helper methods for writing-mode conversions of position-area values.
The core mapping logic is in PositionArea::coordMatchedTrackForAxis().

We can consider adding PositionArea::flowMatchedTrackForAxis() if future
layout code needs it, but for now positioning is handled in RenderBox
coordinates, so that&apos;s all that&apos;s needed at the moment.

* Source/WebCore/platform/BoxSides.h:
(WebCore::WritingMode::blockAxis const): Implement method to return physical axis value.
(WebCore::WritingMode::inlineAxis const): Implement method to return physical axis value.
(WebCore::WritingMode::horizontalAxis const): Implement method to return logical axis value.
(WebCore::WritingMode::verticalAxis const): Implement method to return logical axis value.
* Source/WebCore/platform/text/WritingMode.h: Declare methods to return axis values.
* Source/WebCore/rendering/style/PositionArea.cpp:
(WebCore::PositionArea::coordMatchedTrackForAxis const): Implement writing-mode mapping for position-area values.
* Source/WebCore/rendering/style/PositionArea.h: Make enum values explicit, since we&apos;re encoding information into them.
(WebCore::isPositionAreaAxisLogical): Add helper method.
(WebCore::isPositionAreaDirectionLogical): Add helper method.
(WebCore::mapPositionAreaAxisToPhysicalAxis): Add helper method.
(WebCore::mapPositionAreaAxisToLogicalAxis): Add helper method.
(WebCore::flipPositionAreaTrack): Add helper method.

Canonical link: <a href="https://commits.webkit.org/291042@main">https://commits.webkit.org/291042@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5d191ca130c26381610264617e946805e86f27e0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91726 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11254 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/798 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96681 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42374 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93776 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11625 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19859 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70483 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27920 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94727 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8891 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83099 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50749 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8636 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/691 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41567 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78951 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/692 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98695 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18870 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14019 "Build is in progress. Recent messages:") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79445 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19123 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78943 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78666 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23185 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/529 "Found 1 new test failure: compositing/repaint/become-overlay-composited-layer.html (failure)") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11958 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14561 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18861 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24110 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18564 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22019 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20318 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/BoxSides.h:
```diff
@@ -38,13 +38,13 @@ namespace WebCore {
 enum class LogicalBoxAxis : uint8_t {
-    Inline,
-    Block
+    Inline = 0,
```
Doesn't seem helpful.

fantasai replied:
I don't feel strongly, so fine to leave as-is; but I had made the change to make it more explicit that these values are part of this API, and changing them could affect other code.

cdumez commented on Source/WebCore/platform/BoxSides.h:
```diff
@@ -38,13 +38,13 @@ namespace WebCore {
 enum class BoxAxis : uint8_t {
-    Horizontal,
-    Vertical
+    Horizontal = 0,
```
ditto.

cdumez commented on Source/WebCore/platform/BoxSides.h:
```diff
@@ -119,18 +119,28 @@ constexpr LogicalBoxCorner mapCornerPhysicalToLogical(WritingMode, BoxCorner);
 
+constexpr BoxAxis WritingMode::blockAxis() const
+{
+    return static_cast<BoxAxis>(isHorizontal());
```
A ternary would be less risky / more robust than a static_cast

cdumez replied:
I am no CSS person but I also find it confusing that `isHorizontal() == true` seems to translate to `BoxAxis::Vertical` and not `BoxAxis::Horizontal` 🤷🏻

cdumez replied:
Note that clang generates the same assembly code for both functions here:
```
enum class BoxAxis : bool {
    Vertical,
    Horizontal
};

BoxAxis blockAxis(bool isHorizontal)
{
    return static_cast<BoxAxis>(isHorizontal);
}

BoxAxis blockAxis2(bool isHorizontal)
{
    return isHorizontal ? BoxAxis::Horizontal : BoxAxis::Vertical;
}
```

It is thus possible to have the same perf with a ternary than with a static_cast, without the code being as fragile.

cdumez commented on Source/WebCore/platform/BoxSides.h:
```diff
@@ -119,18 +119,28 @@ constexpr LogicalBoxCorner mapCornerPhysicalToLogical(WritingMode, BoxCorner);
+
+constexpr BoxAxis WritingMode::inlineAxis() const
+{
+    return static_cast<BoxAxis>(!isHorizontal());
```
ditto

cdumez commented on Source/WebCore/platform/BoxSides.h:
```diff
@@ -119,18 +119,28 @@ constexpr LogicalBoxCorner mapCornerPhysicalToLogical(WritingMode, BoxCorner);
     bool isBlock = logicalAxis == LogicalBoxAxis::Block;
     bool isVertical = isBlock != writingMode.isVertical();
-    return isVertical ? BoxAxis::Vertical : BoxAxis::Horizontal;
+    return static_cast<BoxAxis>(isVertical);
```
This is more fragile, what's the reason for this change?

cdumez commented on Source/WebCore/platform/BoxSides.h:
```diff
@@ -119,18 +119,28 @@ constexpr LogicalBoxCorner mapCornerPhysicalToLogical(WritingMode, BoxCorner);
     bool isVertical = axis == BoxAxis::Vertical;
     bool isBlock = isVertical != writingMode.isVertical();
-    return isBlock ? LogicalBoxAxis::Block : LogicalBoxAxis::Inline;
+    return static_cast<LogicalBoxAxis>(isBlock);
```
ditto.

cdumez commented on Source/WebCore/rendering/style/PositionArea.h:
```diff
@@ -25,43 +25,76 @@
+
+// Specifies which tracks(s) on the axis that the position-area span occupies.
+// Represented as 3 bits: start track, center track, end track.
+enum class PositionAreaTrack : uint8_t {
```
Have you considered using a `WTF::OptionSet` instead of doing manual bit handling?

fantasai replied:
This should look, to the outside, like an enum. It's only internal methods like `invertPositionAreaTrack` that need to know about its internal information hacks.

smfr replied:
I tend to agree with Chris, that it would be nice to see in the code how those values are computed. If they are bit masks, you should be able to express that in code.

tuankiet65 commented on Source/WebCore/rendering/style/PositionArea.h:
```diff
@@ -25,43 +25,76 @@
 namespace WebCore {
 
 // The axis that the span specifies.
+// Represented as 3 bits, with the last one indicating the corresponding [Logical]BoxAxis value.
```
I think utilizing each bit to indicate the physical/logical axis/direction is pretty neat. Maybe you should explain what bit 1 and 2 means in this comment too, I only understand what they mean after looking at `isPositionAreaAxisLogical` and `isPositionAreaDirectionLogical`. Something like "bit 2 is physical or logical axes, bit 1 is physical or logical directions"

tuankiet65 commented on Source/WebCore/rendering/style/PositionArea.h:
```diff
@@ -25,43 +25,76 @@
+
+static inline BoxAxis mapPositionAreaAxisToPhysicalAxis(const PositionAreaAxis positionAxis, const WritingMode writingMode)
+{
+    auto physicalAxis = static_cast<uint8_t>(positionAxis) & 0b001;
```
Maybe put this bitwise into a separate function (like `isPositionAreaDirectionLogical`)? So we'd have three functions for each bit, and it's a bit easier to read than a bitwise.

fantasai replied:
I don't want anyone else (outside of these methods that "belong" to PositionAreaAxis) to be extracting out this bit. :) It has no meaning on its own, it needs to be converted through one of these methods. I can make it a named constant within the function though.

tuankiet65 replied:
Right, I mean just like `isPositionAreaAxisLogical` and `isPositionAreaDirectionLogical`, we also have a function to extract that bit. And it'll be `static inline` so it wouldn't be exposed to anything else. 

`mapPositionAreaAxisToPhysicalAxis` and `mapPositionAreaAxisToLogicalAxis` has this logic duplicated also, so might be a good way to de-duplicate it.

fantasai replied:
Yeah, I don't love the duplicated const declarations, but this is a .h file, so even a static method is exposed to anything that imports the enum, right?

tuankiet65 replied:
That's true, I neglect this is a header file. Let's keep thing the current way then.

tuankiet65 commented on Source/WebCore/rendering/style/PositionArea.h:
```diff
@@ -25,43 +25,76 @@
+
+static inline LogicalBoxAxis mapPositionAreaAxisToLogicalAxis(const PositionAreaAxis positionAxis, const WritingMode writingMode)
+{
+    auto logicalAxis = static_cast<uint8_t>(positionAxis) & 0b001;
```
Ditto above.

tuankiet65 commented on Source/WebCore/rendering/style/PositionArea.h:
```diff
@@ -25,43 +25,76 @@
 
-    // Center and last tiles.
-    SpanEnd,
+static inline PositionAreaTrack invertPositionAreaTrack(PositionAreaTrack track)
```
Very clever way to flip the tracks, although it took me a while to see the logic behind all the bitwises 😅 I'd certainly appreciate some more comments here.

tuankiet65 replied:
And I think "flip" is more appropriate than "invert"?

fantasai replied:
Sounds good! Changed it over. Actually, we could even just call it `flip()`, since the type name isn't adding anything here.

tuankiet65 replied:
This is one of those moments where I wish C++ has enum methods so we could have a `PositionAreaTrack::flip` or something like that...

`flipPositionAreaTrack` is too verbose but I feel like `flip` is too ambiguous, so maybe `flipTrack`?

fantasai replied:
I have many of those moments. :)

I think we should either go with `flip()` and rely on the type information, or go with `flipPositionAreaTrack()` and be explicit about it. It's not a commonly used function (there's probably only going to be somewhere between 1-3 calls in our codebase overall) so I don't think we need to worry about the length.

tuankiet65 replied:
+1 for `flipPositionAreaTrack`.

tuankiet65 reviewed (commented):
I don't see anything obviously wrong (yet)

smfr commented on Source/WebCore/platform/BoxSides.h:
```diff
@@ -121,6 +121,26 @@ constexpr LogicalBoxCorner mapCornerPhysicalToLogical(const WritingMode, const B
 
 /** Implementation Below **********************************************/
```
`// Mark`

fantasai replied:
???

nt1m replied:
https://searchfox.org/wubkat/search?q=%2F%2F+MARK%3A+-&path=&case=false&regexp=false

fantasai committed:
Support writing-mode conversions for position-area
https://bugs.webkit.org/show_bug.cgi?id=288261
rdar://145333147

Reviewed by Alan Baradlay.

Adds helper methods for writing-mode conversions of position-area values.
The core mapping logic is in PositionArea::coordMatchedTrackForAxis().

We can consider adding PositionArea::flowMatchedTrackForAxis() if future
layout code needs it, but for now positioning is handled in RenderBox
coordinates, so that's all that's needed at the moment.

* Source/WebCore/platform/BoxSides.h:
(WebCore::WritingMode::blockAxis const): Implement method to return physical axis value.
(WebCore::WritingMode::inlineAxis const): Implement method to return physical axis value.
(WebCore::WritingMode::horizontalAxis const): Implement method to return logical axis value.
(WebCore::WritingMode::verticalAxis const): Implement method to return logical axis value.
* Source/WebCore/platform/text/WritingMode.h: Declare methods to return axis values.
* Source/WebCore/rendering/style/PositionArea.cpp:
(WebCore::PositionArea::coordMatchedTrackForAxis const): Implement writing-mode mapping for position-area values.
* Source/WebCore/rendering/style/PositionArea.h: Make enum values explicit, since we're encoding information into them.
(WebCore::isPositionAreaAxisLogical): Add helper method.
(WebCore::isPositionAreaDirectionLogical): Add helper method.
(WebCore::mapPositionAreaAxisToPhysicalAxis): Add helper method.
(WebCore::mapPositionAreaAxisToLogicalAxis): Add helper method.
(WebCore::flipPositionAreaTrack): Add helper method.

Canonical link: https://commits.webkit.org/291042@main
